common/workunit/workunit.cpp

```diff
@@ -2659,83 +2659,67 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
-    for (it->first(); it->isValid(); )
+    StatsScopeId globalScopeId(SSTglobal, (unsigned)0);
+    statsCollection.setown(createStatisticCollection(globalScopeId));
+
+    class StatsCollectionAggregatesLoader : public IWuScopeVisitor
     {
-        cost_type value = 0;
-        if (it->getStat(StCostFileAccess, value))
-        {
-            totalCost += value;
-            it->nextSibling();
-        }
-        else
+    public:
+        StatsCollectionAggregatesLoader(IStatisticCollection * _statsCollection) : statsCollection(_statsCollection) {}
+
+        virtual void noteStatistic(StatisticKind kind, unsigned __int64 value, IConstWUStatistic & extra) override
         {
-            it->next();
+            statsCollection->setStatistic(extra.queryScope(), kind, value);
         }
-    }
-    return totalCost;
-}
+        virtual void noteAttribute(WuAttr attr, const char * value) override { throwUnexpected(); }
+        virtual void noteHint(const char * kind, const char * value) override { throwUnexpected(); }
+        virtual void noteException(IConstWUException & exception) override { throwUnexpected(); }
+    private:
+        Linked<IStatisticCollection> statsCollection;
+    };
 
-void gatherSpillSize(const IConstWorkUnit * wu, const char *scope, stat_type & peakSizeSpill)
-{
     WuScopeFilter filter;
-    if (!isEmptyString(scope))
-        filter.addScope(scope);
-    else
-    {
-        filter.addScope("");
-        filter.addSource("global");
-    }
-    filter.setIncludeNesting(1);
-    filter.addOutputStatistic(StSizeGraphSpill);
-    filter.addRequiredStat(StSizeGraphSpill);
+    filter.addScopeType(SSTglobal).addScopeType(SSTworkflow).addScopeType(SSTgraph);
+    const unsigned numStats = aggregateKindsMapping.numStatistics();
+    for (unsigned i=0; i<numStats; ++i)
+        filter.addOutputStatistic(aggregateKindsMapping.getKind(i));
+    filter.setDepth(1,3); // 1=global, 2=workflow, 3=graph
+    filter.setSources(SSFsearchGlobalStats);
+    filter.setIncludeNesting(0);
     filter.finishedFilter();
-    Owned<IConstWUScopeIterator> it = &wu->getScopeIterator(filter);
-    peakSizeSpill = 0;
-    for (it->first(); it->isValid(); )
-    {
-        stat_type value = 0;
-        if (it->getStat(StSizeGraphSpill, value))
-        {
-            if (value>peakSizeSpill)
-                peakSizeSpill = value;
-            it->nextSibling();
-        }
-        else
-        {
-            it->next();
-        }
-    }
+
+    StatsCollectionAggregatesLoader aggregatesLoader(statsCollection);
+    Owned<IConstWUScopeIterator> iter = &workunit.getScopeIterator(filter);
+    ForEach(*iter)
+        iter->playProperties(aggregatesLoader);
 }
 
-void updateSpillSize(IWorkUnit * wu, const char * scope, StatisticScopeType scopeType)
+// Replace the stats at the specified scope level
+void StatisticsAggregator::recordStats(IStatisticCollection * sourceStats, unsigned wfid, const char * graphName, unsigned sgId)
```

Comment:
Also, I don't think 'recordStats' needs or should be part of CStatisticCollection.
I think the only methods that are needed in CStatisticCollection so that can be used by the aggregate class are
1) a method that ensures the scope exists.
2) a way to set/check dirty, 
3) and a way to update an [aggregte] stats.

then this method can be something like:

```
void StatisticsAggregator::copyAggregates(IStatisticCollection * sourceStats) 
{
    StringBuffer scope;
    sourceStats->getFullScope(scope);

    IStatisticCollection * targetStats = statsCollection->ensureScope(scope); // create intermediate scopes and returns tail
    bool wasUpdated = false;
    
    // More efficient to iterate over stats rather than mapping...
    unsigned numStats = sourceStats->getNumStatistics();
    for (unsigned s=0; s<numStats; s++)
    {
        StatisticKind kind;
        unsigned __int64 value;
        sourceStats->getStatistic(kind, value, s);
        if (kind != (StatisticKind)(kind & StKindMask))
            continue; // ignore variants
        if (mapping.hasKind(kind))
        {
            if (targetStats->updateStatistic(kind, value, StatsMergeReplace))
                wasUpdated=true;
        }
    }
    if (wasUpdated)
        targetStats->markDirty();
```